openedx/features/enterprise_support/enrollments/utils.py

```diff
@@ -0,0 +1,103 @@
+    current uses of the EnrollmentApiClient
+    The REST enrollment endpoint may also eventually also want to eventually reuse this function
+    since it's a subset of what the endpoint handles
+
+    Unlike the REST endpoint, this function does not check for enterprise enabled, or user api key
+    permissions etc. Those concerns are still going to be used by REST endpoint but this function
+    is meant for use from within edx-enterprise hence already presume such privileges.
+
+    Arguments:
+     - username (str): User name
+     - course_id (obj) : Course key obtained using CourseKey.from_string(course_id_input)
+     - mode (CourseMode): course mode
+     - cohort_name (str): Optional. If provided, user will be added to cohort
+     - is_active (bool): Optional. A Boolean value that indicates whether the
+        enrollment is to be set to inactive (if False). Usually we want a True if enrolling anew.
+
+    Returns: A serializable dictionary of the new course enrollment.
+    """
+    user = _validate_enrollment_inputs(username, course_id)
+
+    with transaction.atomic():
+        try:
+            # Will reactivate inactive enrollments.
+            response = api.add_enrollment(
+                username,
+                str(course_id),
+                mode=mode,
+                is_active=is_active,
+                enrollment_attributes=None
+            )
+            add_user_to_course_cohort(cohort_name, course_id, user)
+            log.info('The user [%s] has been enrolled in course run [%s].', username, course_id)
+            return response
+        except CourseEnrollmentExistsError as error:
+            log.warning('An enrollment already exists for user [%s] in course run [%s].', username, course_id)
+            return None
+        except CourseEnrollmentError as error:
+            log.exception("An error occurred while creating the new course enrollment for user "
+                          "[%s] in course run [%s]", username, course_id)
+            raise error
+        except CourseUserGroup.DoesNotExist as error:
+            log.exception('Missing cohort [%s] in course run [%s]', cohort_name, course_id)
+            raise error
+        finally:
+            # Assumes that the ecommerce service uses an API key to authenticate.
+            current_enrollment = api.get_enrollment(username, str(course_id))
+            if current_enrollment:
+                audit_log(
+                    'enrollment_change_requested',
+                    course_id=str(course_id),
+                    requested_mode=mode,
+                    actual_mode=current_enrollment['mode'] if current_enrollment else None,
+                    requested_activation=is_active,
+                    actual_activation=current_enrollment['is_active'] if current_enrollment else None,
+                    user_id=user.id
+                )
+
+
+def _validate_enrollment_inputs(username, course_id):
+    """
+    Validates username and course_id.
```

Comment:
Something about this method bothers me, but I can't articulate why. I know it's because we're not really checking anything with course_id except that it was passed into this method or not, but I see the use of that in the situation that `lms_enroll_user_in_course` is provided a second argument that's a variable that equates to none. 
ie:
`lms_enroll_user_in_course('userid')` will fail immediately before even hitting this validation while
```
some_var = None
lms_enroll_user_in_course('userid', some_var)
```
will not, thus we need to catch that error ahead of time in validation as you're doing here.

But something about it equating to the user and also returning a user object instead of just validating bugs me  🤔 
Anyways, for now this seems totally fine though. Certainly gets the job done and combines two similar error conditions.

Reply:
thanks yeah I think written more cleanly it should actually just return/store cleaned outputs for use in rest of the code. Something like what serializers validate_inputs do. 

I fetched user so we did not have to refetch after checking.  but we can also just let it fail openly if User is not found (enrollment apis later will also fail probably anyway). We can surely revisit later if something acts up here.